File: app/src/main/java/me/eigenein/nexttrainwear/fragments/TrainsFragment.java
package me.eigenein.nexttrainwear.fragments;

import android.app.Fragment;
import android.content.Context;
import android.location.Location;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.LinearSnapHelper;
import android.support.wearable.view.WearableRecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.location.LocationServices;

import java.util.Set;

import me.eigenein.nexttrainwear.Preferences;
import me.eigenein.nexttrainwear.R;
import me.eigenein.nexttrainwear.Station;
import me.eigenein.nexttrainwear.StationCatalogue;
import me.eigenein.nexttrainwear.Utils;

public class TrainsFragment
    extends Fragment
    implements GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener {

    final private static String TAG = TrainsFragment.class.getSimpleName();
    final private static Station DEFAULT_STATION = StationCatalogue.STATION_BY_CODE.get("ASD");

    public static TrainsFragment newInstance() {
        return new TrainsFragment();
    }

    private GoogleApiClient apiClient;

    public TrainsFragment() {
        // Do nothing.
    }

    @Override
    public View onCreateView(
        final LayoutInflater inflater,
        final ViewGroup container,
        final Bundle savedInstanceState
    ) {
        final View view = inflater.inflate(R.layout.fragment_trains, container, false);

        final WearableRecyclerView stationsRecyclerView =
            (WearableRecyclerView)view.findViewById(R.id.fragment_trains_stations_recycler_view);
        stationsRecyclerView.setLayoutManager(
            new LinearLayoutManager(getActivity(), LinearLayoutManager.HORIZONTAL, false));
        new LinearSnapHelper().attachToRecyclerView(stationsRecyclerView);

        return view;
    }

    @Override
    public void onAttach(final Context context) {
        super.onAttach(context);
        apiClient = new GoogleApiClient.Builder(context)
            .addApi(LocationServices.API)
            .addConnectionCallbacks(this)
            .addOnConnectionFailedListener(this)
            .build();
    }

    @Override
    public void onResume() {
        super.onResume();
        apiClient.connect();
    }

    @Override
    public void onPause() {
        super.onPause();
        apiClient.disconnect();
    }

    @Override
    public void onConnected(@Nullable final Bundle bundle) {
        try {
            final Location location = LocationServices.FusedLocationApi.getLastLocation(apiClient);
            if (location != null) {
                Log.d(TAG, "Found location: " + location);
                onStationDetected(Utils.INSTANCE.getNearestStation(location));
            } else {
                Log.e(TAG, "Missing last known location");
                onStationDetected(null);
            }
        } catch (final SecurityException e) {
            Log.e(TAG, "Forbidden to obtain last known location", e);
            onStationDetected(null);
        }
    }

    @Override
    public void onConnectionSuspended(final int i) {
        // Do nothing.
    }

    @Override
    public void onConnectionFailed(@NonNull final ConnectionResult connectionResult) {
        Log.e(TAG, "Connection failed: " + connectionResult);
        onStationDetected(null);
    }

    private void onStationDetected(Station station) {
        Log.d(TAG, "Detected station: " + station);
        if (station == null) {
            Log.w(TAG, "Using default station");
            station = DEFAULT_STATION;
        }

        final Set<String> favoriteStations = Preferences.INSTANCE.getFavoriteStations(getActivity());
        favoriteStations.remove(station.getCode()); // don't go to the current station
        // TODO: sort stations by distance from the current station.
        // TODO: empty station set (pick some nearest ones).
        // TODO: destinations adapter.
        // TODO: ride adapter.
    }
}

File: app/src/main/java/me/eigenein/nexttrainwear/fragments/TrainsFragment.kt
package me.eigenein.nexttrainwear.fragments

import android.app.Fragment
import android.os.Bundle
import android.os.Handler
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.LinearSnapHelper
import android.support.wearable.view.WearableRecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationServices
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import me.eigenein.nexttrainwear.R
import me.eigenein.nexttrainwear.adapters.JourneyOptionsAdapter
import me.eigenein.nexttrainwear.adapters.RoutesAdapter
import me.eigenein.nexttrainwear.data.DetectedStation
import me.eigenein.nexttrainwear.data.Station
import me.eigenein.nexttrainwear.data.Stations
import me.eigenein.nexttrainwear.utils.*
import java.util.concurrent.TimeUnit

class TrainsFragment : Fragment() {

    private val handler = Handler()
    private val disposable = CompositeDisposable()
    private val adapter = RoutesAdapter()

    private lateinit var progressLayout: View
    private lateinit var routesRecyclerView: WearableRecyclerView

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val view = inflater.inflate(R.layout.fragment_trains, container, false)

        progressLayout = view.findViewById(R.id.fragment_trains_progress_layout)

        routesRecyclerView = view.findViewById(R.id.fragment_trains_recycler_view)
        routesRecyclerView.layoutManager = LinearLayoutManager(activity, LinearLayoutManager.HORIZONTAL, false)
        routesRecyclerView.adapter = adapter
        routesRecyclerView.setHasFixedSize(true)
        LinearSnapHelper().attachToRecyclerView(routesRecyclerView)

        return view
    }

    override fun onResume() {
        super.onResume()

        // Obtain current location and build possible route list.
        val resumeTime = System.currentTimeMillis()
        disposable.add(
            GoogleApiClient.Builder(activity)
                .addApi(LocationServices.API)
                .asFlowable()
                .flatMap {
                    LocationRequest.create()
                        .setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY)
                        .asFlowable(it)
                        .filter { it.time >= resumeTime } // we only want fresh location fixes
                        .take(1) // otherwise we'll get timeout error anyway
                        .timeout(LOCATION_TIMEOUT_SECONDS, TimeUnit.SECONDS)
                }
                .map { DetectedStation(true, Station.findNearestStation(it)) }
                .doOnError { Log.e(LOG_TAG, "Failed to detect station: " + it) }
                .onErrorReturn {
                    val lastStation = Stations.STATION_BY_CODE[Preferences.getLastStationCode(activity)]
                    DetectedStation(false, lastStation ?: Stations.AMSTERDAM_CENTRAAL)
                }
                .subscribeOn(AndroidSchedulers.mainThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { onStationDetected(it) }
        )

        // Refresh countdown and auto-scroll journey options when train departs.
        disposable.add(
            handler.asFlowable(COUNTDOWN_UPDATE_INTERVAL_MILLIS).subscribe {
                val journeyOptionsRecyclerView = routesRecyclerView
                    .findFirstVisibleViewHolder<RoutesAdapter.ViewHolder>()
                    ?.journeyOptionsRecyclerView
                journeyOptionsRecyclerView
                    ?.findVisibleViewHolders<JourneyOptionsAdapter.ViewHolder>()
                    ?.forEach {
                        if (it.refreshCountDown() in AUTO_SCROLL_THRESHOLD_MILLIS..0 && journeyOptionsRecyclerView.isScrollIdle()) {
                            // Departed. Scroll to the next one.
                            journeyOptionsRecyclerView.smoothScrollToPosition(it.adapterPosition + 1)
                            @Suppress("DEPRECATION")
                            activity.getVibrator().vibrate(AUTO_SCROLL_VIBRATE_PATTERN, -1)
                        }
                    }
            }
        )

        // Refresh journey options.
        disposable.add(
            handler.asFlowable(JOURNEY_OPTIONS_REFRESH_INTERVAL_MILLIS).subscribe {
                // FIXME: this can lead to a pair of concurrent requests.
                routesRecyclerView.findFirstVisibleViewHolder<RoutesAdapter.ViewHolder>()?.refreshJourneyOptions()
            }
        )
    }

    override fun onPause() {
        super.onPause()
        disposable.clear()
    }

    /**
     * Updates the fragment based on the detected station.
     */
    private fun onStationDetected(detectedStation: DetectedStation) {
        Log.d(LOG_TAG, "Departure station: " + detectedStation)
        Preferences.setLastStationCode(activity, detectedStation.station.code)

        val destinations = selectDestinations(detectedStation.station)
        Log.d(LOG_TAG, "Selected destinations: " + destinations.size)

        // Show routes view.
        progressLayout.visibility = View.GONE
        routesRecyclerView.visibility = View.VISIBLE
        adapter.swap(
            detectedStation.usingLocation,
            destinations.map { detectedStation.station.routeTo(it) }
        )
    }

    /**
     * Select destination stations to go to from the specified departure station.
     */
    private fun selectDestinations(departureStation: Station): List<Station> {
        val stationCodes = Preferences.getStations(activity)

        // Select favorite stations sorted by distance.
        val favoriteStations = stationCodes
            .filter { it != departureStation.code }
            .mapNotNull { Stations.STATION_BY_CODE[it] }
            .sortedBy { departureStation.distanceTo(it.latitude, it.longitude) }

        // Select some other stations sorted by distance from current.
        val allStations = Stations.ALL_STATIONS
            .filter { it.code !in stationCodes && it.code != departureStation.code }
            .sortedBy { departureStation.distanceTo(it.latitude, it.longitude) }

        return favoriteStations + allStations
    }

    companion object {

        private const val LOCATION_TIMEOUT_SECONDS = 5L // FIXME: exponetial backoff.
        private const val COUNTDOWN_UPDATE_INTERVAL_MILLIS = 1000L
        private const val JOURNEY_OPTIONS_REFRESH_INTERVAL_MILLIS = 60000L
        private const val AUTO_SCROLL_THRESHOLD_MILLIS = -1500L // FIXME: better ideas?

        private val AUTO_SCROLL_VIBRATE_PATTERN = longArrayOf(0L, 200L, 200L, 200L)
        private val LOG_TAG = TrainsFragment::class.java.simpleName
    }
}
